### PR TITLE
feat: light-godwoken-cli tool

### PR DESCRIPTION
**Pull request by ShookLyngs**
## Light-Godwoken CLI
I am working on a `light-godwoken-cli` tool that will allow us to interact with Godwoken easier.

For more about the tool:
https://github.com/nervosnetwork/godwoken-tests/blob/b2a85a3c6ef6ee54580b8eccee12d71fbef9d8c1/scripts/light-godwoken-cli/README.md

## Status
The `light-godwoken-cli` tool can to be used in the following tatks:
- [x] Withdrawal testing in `reusable-integration-test-v1`
  - [x] Construct `LightGodwokenConfigMap` for Devnet_v1
  - [x] Add this test into the workflow
- [ ] Test workflows for looping deposit/withdrawal
  - [x] Construct `LightGodwokenConfigMap` for Alphanet_v1
  - [x] Do we need abilities for deposit/withdrawal SUDT?
  - [ ] ~~Construct workflow~~ (leave for another pull)

### 1. Network configs for light-godwoken-cli
Currently, we can use this tool for deposit/withdrawl transactions on `Testnet_v1` (Should work on `Mainnet_v1` too, I haven't test it yet), but on `Devnet_v1` and `Alphanet_v1` it won't work.

The reason behind this is because light-godwoken has a set of predefined configs for `Testnet_v1` and `Mainnet_v1`:
https://github.com/nervosnetwork/light-godwoken/blob/develop/packages/light-godwoken/src/constants/lightGodwokenConfig.ts

In order for the tool to work on `Devnet_v1` and `Alphanet_v1`, we need to construct a set of configs for those networks. For the configs' construction, those fileds in configs are overly complex for me, I need help from @Flouse.

### 2. Looping test workflow
Currently the cli tool only supports deposit/withdrawal CKB and not SUDT, I'm not sure if the cli tool can satisfy the `looping deposit/withdrawal` workflow. 

So for this problem we might need more discussions.

### CODE REVIEW

File: scripts/light-godwoken-cli/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Light-Godwoken CLI
+
+A CLI tool to interact with LightGodwoken.
+
+## Installation
+
+### Setup light-godwoken submodule
+```shell
+# position: godwoken-tests/
+$ git submodule update --init
```
Comment (Flouse) [inline scripts/light-godwoken-cli/README.md]:
```suggestion
$ git submodule update --init --recursive --depth=1
```

File: scripts/light-godwoken-cli/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+### Deposit
+Deposit capacity from CKB layer1 to Godwoken layer2:
+```shell
+# position: godwoken-tests/scripts/light-godwoken-cli/
+$ lgc deposit -p <PRIVATE_KEY> -c <DEPOSIT_CAPACITY> -n <NETWORK>
+```
+
+### Withdrawal
+Withdrawal capacity from Godwoken layer2 to CKB layer1:
```
Comment (Flouse) [inline scripts/light-godwoken-cli/README.md]:
```suggestion
Withdraw capacity from Godwoken layer2 to CKB layer1:
```

**Review by Flouse — CHANGES_REQUESTED** [review]

**Comment by Flouse** [comment]
> * Do we need abilities for deposit/withdrawal SUDT?

yes, this could be implemented in another PR.

**Comment by Flouse** [comment]
> Currently, we can use this tool for deposit/withdrawl transactions on `Testnet_v1` (Should work on `Mainnet_v1` too, I haven't test it yet), but on `Devnet_v1` and `Alphanet_v1` it won't work.
> 
> The reason behind this is because light-godwoken has a set of predefined configs for `Testnet_v1` and `Mainnet_v1`: https://github.com/nervosnetwork/light-godwoken/blob/develop/packages/light-godwoken/src/constants/lightGodwokenConfig.ts
> 
> In order for the tool to work on `Devnet_v1` and `Alphanet_v1`, we need to construct a set of configs for those networks. For the configs' construction, those fileds in configs are overly complex for me, I need help from @Flouse.

Shall we use `poly_version` RPC to get the configs of a network?
```bash
curl https://godwoken-alphanet-v1.ckbapp.dev -X POST \
  -H "Content-Type: application/json" \
  -d '{"jsonrpc": "2.0", "method":"poly_version", "params": [], "id": 1}'

# Result
{"jsonrpc":"2.0","id":1,"result":{"versions":{"web3Version":"1.6.1-rc1","web3IndexerVersion":"1.6.1-rc1","godwokenVersion":"1.5.0-rc2 c36b022"},
"nodeInfo": {
  "nodeMode": "fullnode",
  "rollupCell": {
    "typeHash": "0x0a6d3bb9392242c0357fcae7ccead8f87b428ed280909bfdfaa549529b85df38",
    "typeScript": {
      "code_hash": "0x56abab7961e8348aed629a0e59c05d0f6b555314f8f95606eae4bcb2adafdce3",
      "hash_type": "type",
      "args": "0x749f79c58129fb18ce425e030e23f127fe60979ef8f69c28a945f4da19fec591"
    }
  },
  "rollupConfig": {
    "requiredStakingCapacity": "0x2d79883d2000",
    "challengeMaturityBlocks": "0x1c2",
    "finalityBlocks": "0x64",
    "rewardBurnRate": "0x32",
    "chainId": "0x315de"
  },
  "gwScripts": {
    "deposit": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0xae79482d7b4e639d2ff323465821686344e3e19d0d9223b03297ffb0204fb8a1"
      },
      "typeHash": "0x9aa15280cf2123755516ff93180ab14b66a043562ffd70a0947afe7a12d573e5"
    },
    "withdraw": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0x5d4b16e48cd79b8fcbe5ddcff756e8ebebddd9019d7eb24f5333cc6c26bcaa12"
      },
      "typeHash": "0xb580c581b8789dbd3add6c8d2efc63af3590b1a272d950e7dd1c2f64f5b425b5"
    },
    "stateValidator": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0x514a98ddd783a10ef84025e2888b0d2ce22d040d9d0e0b2fa85cea3436d7dbe6"
      },
      "typeHash": "0x56abab7961e8348aed629a0e59c05d0f6b555314f8f95606eae4bcb2adafdce3"
    },
    "stakeLock": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0xce25c4130db0c7b3ab2c5c5a5e5c73f1f0e682387d0dddb7509be7dabd168c06"
      },
      "typeHash": "0xfeac11dee7c66e54e4864c177215d0cbb6ebcfbd2fe8e0d2bc45b3413542e2da"
    },
    "custodianLock": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0xe0a9466ceec538c94e92edc9759a68a2b38f48348a69de920e2841a0239e0497"
      },
      "typeHash": "0x38e7fa830764ad269f63fa9e874b39d13d8f4ca42c47cac20e11c847dc5595bc"
    },
    "challengeLock": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0x057e28dce74a7e6d03e4a3cdc90ec4d096c38ebea880dd6b0d6f549003bb6bc7"
      },
      "typeHash": "0xb580c581b8789dbd3add6c8d2efc63af3590b1a272d950e7dd1c2f64f5b425b5"
    },
    "l1Sudt": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0x4db75e03349f4f2ec792476035dd1b7376c683130f7e2e74024be2d9ee064511"
      },
      "typeHash": "0xc5e5dcf215925f7ef4dfaf5f4b4f105bc321c02776d6e7d52a1db3fcd9d011a4"
    },
    "l2Sudt": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0xc1fc7c7a3c09ed5e637128fadf60a4d87a534011a0773e83f03ac6ff56c329ff"
      },
      "typeHash": "0x6900864b01b49abad8c683cbf8641e3df2c12e362596eb8bab88ce99ff3b6f92"
    },
    "omniLock": {
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0x1b8572b16c07f46a0efed623aea6de05d45985b9a7c1b0b52276da5d9f9615b7"
      },
      "typeHash": "0x79f90bb5e892d80dd213439eeab551120eb417678824f282b4ffb5f21bad2e1e"
    }
  },
  "eoaScripts": {
    "eth": {
      "typeHash": "0x8d6af470fd57d1afe543751eed58afef310ec659d565319f448bcc765176dcdc",
      "script": {
        "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
        "hash_type": "type",
        "args": "0x414e3a6b031cd3c18333fd071d9dd15666233b3889f45c5e9186492daf7fd8f0"
      }
    }
  },
  "backends": {
    "sudt": {
      "validatorCodeHash": "0xe9f21f08e1a09748b701be259c17e1dd9f1accaee23e2c6e0609916d308a1b4d",
      "generatorCodeHash": "0x34f8d804e2173c076495e78458ab9031c769c36062fede0cb35e7762db04224d",
      "validatorScriptTypeHash": "0x6900864b01b49abad8c683cbf8641e3df2c12e362596eb8bab88ce99ff3b6f92"
    },
    "meta": {
      "validatorCodeHash": "0x65c15510932d7ca6622e7a9c84febc4861d50423b4c1dbeb1bc33ad79312dd57",
      "generatorCodeHash": "0xbf9a4e7cdb07f0234efc5ad0631d3797badf0cc7ce6d7d938388a855e9f7eb3f",
      "validatorScriptTypeHash": "0x953a0992e397c76eb0e1cd16ff8998a59c00f9280586fda31e3fe1d99bd6707a"
    },
    "polyjuice": {
      "validatorCodeHash": "0x320f87eb51c6f4bbcbcbed5e31e97e8c4580d49965259519476ec2985b7fa0b6",
      "generatorCodeHash": "0x15b4b918955bb92f036abfc350b9587d1983fba8f94de85973a4c0e70f98b1c1",
      "validatorScriptTypeHash": "0x7907a044582d425a584bd2759ff1855dfc66f0c5391e1be932c9716f58742b15"
    },
    "ethAddrReg": {
      "validatorCodeHash": "0x57553b7cb7736307865ace9250c492460f12291e685d81e8d0a2e033e5582a00",
      "generatorCodeHash": "0x96ac862b9791c0d1cb1ccf85c1f1d60b1e359a1288d52fad6987b9fc53b0b289",
      "validatorScriptTypeHash": "0xdd2c6955089f34785ea049d62335cb06982e665d73687626e6b4af09b8a7774b"
    }
  },
  "accounts": {
    "polyjuiceCreator": {
      "id": "0x4",
      "scriptHash": "0x6e55e1014b011dfa2aa12f4833dcf347d966c5eb5106196f52f945fd048b73a2"
    },
    "ethAddrReg": {
      "id": "0x2",
      "scriptHash": "0xc88f2395683ed4a69141d76f8fc6c612d10f7d8c964f7ddcd0b2182fda545d73"
    },
    "defaultFrom": {
      "id": "0x3",
      "scriptHash": "0x130c25347beb6744ea1c9eb75a6a2b8dd23c04f83402cfdeb81f41915d968544"
    }
  },
  "chainId": "0x315de"
}}}
```

You may check the configs in this Repo:
https://github.com/nervosnetwork/godwoken-info

e.g. https://github.com/nervosnetwork/godwoken-info/tree/main/testnet_v1_1

**Comment by ShookLyngs** [comment]
I think the PR is ready for review.

### Changes: 
- **Better module readability**
- **Provides handy commands**
  - `lgc deposit` -  deposit capacity from L1 to L2
  - `lgc withdraw` - withdraw capacity from L2 to L1
  - `lgc get-balance` - check L1/L2 balance for an account
- **Network supports (only v1 networks):**
  - mainnet_v1
  - testnet_v1
  - alphanet_v1
  - devnet_v1
- **Workflow changes**
  - adds `withdraw v1 test` in `reusable-integration-test-v1.yml` (need review and actually run it)

### Cancels
- **The looping batch-deposit test workflow will leave for another PR**
It's a bit complicated, need further research.
Mainly because there should be a main account, and before running batch workflows, we have to transfer L1 capacity from main account to sub accounts, the the faucet process and the construction of the transaction will need further research.
As to the deposit/withdrawal process , they're already useable in most ways (of course, need review).

**Comment by ShookLyngs** [comment]
I think the PR is ready for review.

### Changes: 
- **Provides handy commands**
  - `lgc deposit` -  deposit capacity from L1 to L2
  - `lgc withdraw` - withdraw capacity from L2 to L1
  - `lgc get-balance` - check L1/L2 balance for an account
- **Network supports (only v1 networks):**
  - mainnet_v1
  - testnet_v1
  - alphanet_v1
  - devnet_v1
- **Workflow changes**
  - adds `withdraw v1 test` in `reusable-integration-test-v1.yml` (need review and actually run it)

### The looping batch-deposit test workflow will leave for another PR
The task is a bit complicated and need further research, mainly because of the extra steps before depositing. 
As to the deposit/withdrawal process , they're already useable in most cases (still need review).

Before running the batch workflows, we should:
1. claim faucet for the main account
2. transfer L1 capacity from the main account to sub accounts 

The construction of L1 transaction will need further research. And besides, it seems that only claim faucet for the main account is not enough, since there could have dozens of sub accounts requires L1 capacity (each deposit transaction have a minimum capacity limit). 

So there's also a solution No.2:
1. only claim faucet for every sub accounts

It will waste many capacity from the faucet pool, but it is a simple solution since it will save me from manually building L1 transactions.

**Comment by keroro520** [comment]
> Shall we use `poly_version` RPC to get the configs of a network?

Because that [Lumos config includes the dep-cell's out-point](https://github.com/nervosnetwork/light-godwoken/blob/develop/packages/light-godwoken/src/constants/lightGodwokenConfig.ts#L8-L9), but poly_version does not provide it, it can't construct the complete Lumos configuration.

Out points information is stored in [scripts-deploy-result.json](https://github.com/nervosnetwork/godwoken-info/blob/main/testnet_v1_1/scripts-deploy-result.json)

**Review by keroro520 — APPROVED** [review]
AWESOME 👍

**Comment by keroro520** [comment]
Is there any plan to make light-godwoken-cli standalone?

**Comment by ShookLyngs** [comment]
> Is there any plan to make light-godwoken-cli standalone?

No not yet, because the whole light-godwoken toolchain is not ready for it.

**Comment by ShookLyngs** [comment]
@keroro520 

By the way, the tool actually has a problem not yet solved, do you think maybe you can help? 
It's not urgent tho.

The config of `lumos` has a `ANYONE_CAN_PAY` lock, and I have no way finding its info on devnet.
I have filled every other config fields in the devnet's config file, but I still don't know if I'm doing it right.

Here the highlighted script is wrong (copied from another network's config):
https://github.com/nervosnetwork/godwoken-tests/blob/b2a85a3c6ef6ee54580b8eccee12d71fbef9d8c1/scripts/light-godwoken-cli/src/configs/devnet.ts#L37-L44

**Comment by keroro520** [comment]
> @keroro520
> 
> By the way, the tool actually has a problem not yet solved, do you think maybe you can help? It's not urgent tho.
> 
> The config of `lumos` has a `ANYONE_CAN_PAY` lock, and I have no way finding it's info on devnet. I have filled every other config fields in the devnet's config file, but I still don't know if I'm doing it right.
> 
> Here the highlighted script is wrong (copied from another network's config):
> 
> https://github.com/nervosnetwork/godwoken-tests/blob/b2a85a3c6ef6ee54580b8eccee12d71fbef9d8c1/scripts/light-godwoken-cli/src/configs/devnet.ts#L37-L44

What about filling `TX_HASH: "0x0000000000000000000000000000000000000000000000000000000000000000"`, (more explicitly) indicates it is a fake out-point on devnet.

**Review by Flouse — APPROVED** [review]

**Pull request merged**